app/services/definitions.py

```diff
@@ -0,0 +1,543 @@
+    # Get all created schema ids that match the filter
+    bound_logger.debug("Fetching created schemas")
+    response = await handle_acapy_call(
+        logger=bound_logger,
+        acapy_call=aries_controller.schema.get_created_schemas,
+        schema_id=schema_id,
+        schema_issuer_did=schema_issuer_did,
+        schema_name=schema_name,
+        schema_version=schema_version,
+    )
+
+    # Initiate retrieving all schemas
+    schema_ids = response.schema_ids or []
+
+    bound_logger.debug("Getting schemas associated with fetched ids")
+    schemas = await schema_futures(schema_ids, aries_controller)
+
+    return schemas
+
+
+async def schema_futures(
+    schema_ids: List[str], aries_controller: AcaPyClient
+) -> List[CredentialSchema]:
+    """
+    Get schemas with attributes from schema ids
+    """
+    logger.debug("Fetching schemas from schema ids")
+    # We now have schema_ids; the following logic is the same whether called by governance or tenant.
+    # Now fetch relevant schemas from ledger:
+    get_schema_futures = [
+        handle_acapy_call(
+            logger=logger,
+            acapy_call=aries_controller.schema.get_schema,
+            schema_id=schema_id,
+        )
+        for schema_id in schema_ids
+    ]
+
+    # Wait for completion of retrieval and transform all schemas into response model (if a schema was returned)
+    if get_schema_futures:
+        logger.debug("Fetching each of the created schemas")
+        schema_results: List[SchemaGetResult] = await asyncio.gather(
+            *get_schema_futures
+        )
+    else:
+        logger.debug("No created schema ids returned")
+        schema_results = []
+
+    schemas = [
+        credential_schema_from_acapy(schema.var_schema)
+        for schema in schema_results
+        if schema.var_schema
+    ]
+
+    return schemas
+
+
+class CredDefPublisher:
+    def __init__(self, deps: ServiceDependencies):
+        self.deps = deps
+
+    async def assert_public_did(self):
+        try:
+            self.deps.logger.debug("Asserting client has public DID")
+            public_did = await acapy_wallet.assert_public_did(
+                self.deps.aries_controller
+            )
+        except CloudApiException as e:
+            log_message = f"Asserting public DID failed: {e}"
+
+            if e.status_code == 403:
+                self.deps.logger.info(log_message)
+                client_error_message = (
+                    "Wallet making this request has no public DID. "
+                    "Only issuers with a public DID can make this request."
+                )
+
+            else:
+                self.deps.logger.error(log_message)
+                client_error_message = (
+                    "Something went wrong while asserting if request is from a valid issuer. "
+                    "Please try again."
+                )
+            raise CloudApiException(client_error_message, e.status_code) from e
+        return public_did
```

Comment:
asserting public did should be a util method of sorts, reusable anywhere, not specific to this class